apply_DDT.py

```diff
@@ -0,0 +1,562 @@
+        ax.set_xlabel("m(Z')")
+        # cannot use layout_tight, will cause saving errors
+        plt.savefig("plots/best_bdt_cuts.pdf", bbox_inches='tight')
+        plt.savefig("plots/best_bdt_cuts.png", bbox_inches='tight')
+        plt.close()
+
+        
+
+    # _____________________________________________
+    # Apply the DDT to different signals for one BDT cut value
+
+    if 'sig_mt_single_BDT' in plots :
+
+        # grab signal data
+        sig_cols = [Columns.load(f) for f in glob.glob(sig_files)]
+
+        # make plotting objects
+        fig, ax = plt.subplots(figsize=(10, 8))
+        # Options to make the plot fancier 
+        hep.cms.label(rlabel="(13 TeV)")
+ 
+        # Loop over the mZ values
+        for sig_col in sig_cols:
+            mz = sig_col.metadata['mz']
+         
+            # Signal Column 
+            sig_X, sig_pT, sig_mT, sig_rho, sig_weight = bdt_ddt_inputs(sig_col, lumi, all_features)
+            if verbosity > 0 : print("M(Z') = ", mz, " Events: ", len(sig_X), " weights: ", sig_weight)
+
+            # _____________________________________________
+            # Open the trained models and get the scores
+  
+            xgb_model = xgb.XGBClassifier()
+            xgb_model.load_model(model_file)
+            with time_and_log(f'Calculating xgboost scores for {model_file}...'):
+                sig_score = xgb_model.predict_proba(sig_X)[:,1]
+  
+            # _____________________________________________
+            # Apply the DDT  to the signal
+            if verbosity > 0 : print("Applying the DDT signal")
+            sig_score_ddt = calculate_varDDT(sig_mT, sig_pT, sig_rho, sig_score, sig_weight, sig_bdt_cut, ddt_map_file)
+ 
+            # Make mT distributions
+            if mz != 300: alpha = 0.3
+            else: alpha = 1.0
+            score_mask = sig_score_ddt > 0.0 # DDT score > 0.0 is equivalent to BDT score about BDT cut value
+            ax.hist(sig_mT[score_mask], weights=sig_weight[score_mask], bins=50, range=(180,650), histtype='step', label=f"m(Z')={mz}", alpha=alpha)
+ 
+        ax.set_xlabel('$\\mathrm{m}_{\\mathrm{T}}$')
+        ax.set_ylabel('Events')
+        ax.legend()
+        save_plot(plt,'sig_events_vs_mT')
+        
+        # log scale it
+        ax.set_yscale('log')
+        save_plot(plt,'log_sig_events_vs_mT')
+        plt.close()
+
+    # _____________________________________________
+    # Apply the DDT to one signals for different BDT cut values
+
+    if 'one_sig_mt_many_bdt' in plots :
+
+        # grab signal data
+        sig_cols = [Columns.load(f) for f in glob.glob(sig_files)]
+
+        # Loop over the mZ values and only grab the mZ = 300 value
+        sig_col = None
+        for col in sig_cols:
+            mz = col.metadata['mz']
+            if mz == 300 : sig_col = col
+        if sig_col == None : print("The mZ' 300 file doesn't exist, please make sure you provide it in order to make the one sig plots")
```

Comment:
don't be afraid to issue exceptions when the code should stop running because of wrong inputs

Reply:
Done, needs testing